core/src/main/java/com/linecorp/armeria/server/annotation/HttpFileResponseConverterFunction.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.annotation;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Subscriber;
+
+import com.linecorp.armeria.common.FilteredHttpResponse;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpObject;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.file.HttpFile;
+
+/**
+ * A response converter implementation which creates an {@link HttpResponse} when the {@code result} is
+ * an instance of {@link HttpFile}.
+ */
+public final class HttpFileResponseConverterFunction implements ResponseConverterFunction {
+
+    @Override
+    public HttpResponse convertResponse(ServiceRequestContext ctx,
+                                        ResponseHeaders headers,
+                                        @Nullable Object result,
+                                        HttpHeaders trailers) throws Exception {
+        if (result instanceof HttpFile) {
+            final HttpResponse delegate = ((HttpFile) result).asService().serve(ctx, ctx.request());
```

Comment:
How about just returning the `delegate` when `trailers` is empty and `headers` only contains the default `:status`?
You can find the logic about default status [here].(https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java#L198)

Reply:
I've done some debugging and found out that default `:status` header is unknown within this method as we can't use reflection here.

And though we happened to know the default `:status`, we are not able to return `delegate` as it is because `result.asService().serve(...)` will generate a `404` or `200` response which can be different from the default `:status`.

Maybe we can return `delegate` when `trailers` is empty and `headers` only contains `HttpStatus.OK`?
But there will still be an edge case that we can't force the status code to 200 even with `@StatusCode(200)`

If we want `@StatusCode` annotation always overrides the original value, we got to stick the the current implementation I guess.

Reply:
I believe we have to respect the response headers emitted by `HttpFile` service, because otherwise we may be breaking the response; imagine the case where the response has 304 Not Modified and a user specified `@StatusCode(200)`.

How about doing the following:

- Return `delegate` as-is only if:
  - `headers` is empty or contains only `:status` AND `trailers` is empty.
- Return the filtered `delegate` otherwise.
  - When filtering, do not overwrite the headers set by `HttpFile` service. We could do: `headers.toBuilder().set((ResponseHeaders) obj)`, so that a user cannot override the headers set by `HttpFile`.

..?

Reply:
@minwoox @trustin Thanks for the advice! I've updated the code as per your suggestions.

Reply:
> I've done some debugging and found out that default :status header is unknown within this method as we can't use reflection here.

Sorry about making you get confused. 😉 
I was just pointing it to how we set the default status. We know the return of this method which is always `HttpFile`, we can just know that it's `200` if there's no `@StatusCode`.
So I thought if there's the status code which is not `200`, I thought we need to override it.

But as @trustin mentioned maybe it's not a good idea to override the status. 😄